swift3/controllers/multi_upload.py

```diff
@@ -684,6 +693,12 @@ def POST(self, req):
             LOGGER.error(e)
             raise
 
+        s3_etag = '%s-%d' % (s3_etag_hasher.hexdigest(), len(manifest))
```

Comment:
Note: for example, we create etag with "xxx-5" here but at l.705 the last segment can be removed if empty.
I suppose it is safe because etag is saved in sysmeta

Reply:
I just compared with Amazon S3. The empty part is taken into account in the etag.
```
$ aws --profile amazon s3api create-multipart-upload --bucket openio-aymeric --key mpu
{
    "Bucket": "openio-aymeric", 
    "UploadId": "molqZ.jHH2.YbDRgBQOzF1Owlon5YUUALewQWddn4HYjkfZZAqlVYsCF3RjNQKNqQMzcxZfvfZm6.bNs.mbfRjUInIachEIgXr6hYxxbWI9ZfoTo6RpSywUxwBqSfN6o", 
    "Key": "mpu"
}
$ aws --profile amazon s3api upload-part --bucket openio-aymeric --key mpu --part-number 1 --upload-id "molqZ.jHH2.YbDRgBQOzF1Owlon5YUUALewQWddn4HYjkfZZAqlVYsCF3RjNQKNqQMzcxZfvfZm6.bNs.mbfRjUInIachEIgXr6hYxxbWI9ZfoTo6RpSywUxwBqSfN6o" --body part 
{
    "ETag": "\"a65408dc6ff453db90d4e1c7b7e9839c\""
}
$ aws --profile amazon s3api upload-part --bucket openio-aymeric --key mpu --part-number 2 --upload-id "molqZ.jHH2.YbDRgBQOzF1Owlon5YUUALewQWddn4HYjkfZZAqlVYsCF3RjNQKNqQMzcxZfvfZm6.bNs.mbfRjUInIachEIgXr6hYxxbWI9ZfoTo6RpSywUxwBqSfN6o" --body empty-part
{
    "ETag": "\"d41d8cd98f00b204e9800998ecf8427e\""
}
$ aws --profile amazon s3api list-parts --bucket openio-aymeric --key mpu --upload-id "molqZ.jHH2.YbDRgBQOzF1Owlon5YUUALewQWddn4HYjkfZZAqlVYsCF3RjNQKNqQMzcxZfvfZm6.bNs.mbfRjUInIachEIgXr6hYxxbWI9ZfoTo6RpSywUxwBqSfN6o" > manifest
$ aws --profile amazon s3api head-object --bucket openio-aymeric --key mpu
{
    "AcceptRanges": "bytes", 
    "ContentType": "binary/octet-stream", 
    "LastModified": "Mon, 20 Jul 2020 13:18:17 GMT", 
    "ContentLength": 44040192, 
    "ETag": "\"e950edfaf75f03b31136fc64e680266b-2\"", 
    "Metadata": {}
}
```

Reply:
On the other hand, I have a doubt on the usefulness of the code of line 705. This piece of code disappeared with the passage to s3api.
- swift3: https://github.com/openstack-archive/swift3/blob/90db5d1510b2a770387961e7bf0fbeae8101ba45/swift3/controllers/multi_upload.py#L598
- swift: https://github.com/openstack/swift/blob/636b922f3b2882f7dd6c656983d7862b274dcf98/swift/common/middleware/s3api/controllers/multi_upload.py#L612